x-pack/plugins/security_solution/public/onboarding/components/onboarding_body/cards/assistant/components/connectors/connector_setup.tsx

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useState } from 'react';
+import {
+  useEuiTheme,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiListGroup,
+  EuiIcon,
+  EuiPanel,
+  EuiLoadingSpinner,
+  EuiText,
+  EuiLink,
+  EuiTextColor,
+} from '@elastic/eui';
+import { css } from '@emotion/css';
+import {
+  ConnectorAddModal,
+  type ActionConnector,
+} from '@kbn/triggers-actions-ui-plugin/public/common/constants';
+import type { ActionType } from '@kbn/actions-plugin/common';
+import { useKibana } from '../../../../../../../common/lib/kibana';
+import { useFilteredActionTypes } from './hooks/use_load_action_types';
+
+const usePanelCss = () => {
+  const { euiTheme } = useEuiTheme();
+  return css`
+    .connectorSelectorPanel {
+      height: 160px;
+      &.euiPanel:hover {
+        background-color: ${euiTheme.colors.lightestShade};
+      }
+    }
+  `;
+};
+
+interface ConnectorSetupProps {
+  onConnectorSaved?: (savedAction: ActionConnector) => void;
+  onClose?: () => void;
+  compressed?: boolean;
+}
+export const ConnectorSetup = React.memo<ConnectorSetupProps>(
+  ({ onConnectorSaved, onClose, compressed = false }) => {
+    const panelCss = usePanelCss();
+    const {
+      http,
+      triggersActionsUi: { actionTypeRegistry },
+      notifications: { toasts },
+    } = useKibana().services;
+    const [selectedActionType, setSelectedActionType] = useState<ActionType | null>(null);
+
+    const onModalClose = useCallback(() => {
+      setSelectedActionType(null);
+      onClose?.();
+    }, [onClose]);
+
+    const actionTypes = useFilteredActionTypes(http, toasts);
+
+    if (!actionTypes) {
+      return <EuiLoadingSpinner />;
+    }
+
+    return (
+      <>
+        {compressed ? (
+          <EuiListGroup
+            flush
+            data-test-subj="connectorSetupCompressed"
+            listItems={actionTypes.map((actionType) => ({
+              id: actionType.id,
+              label: actionType.name,
+              size: 's',
+              icon: (
+                <EuiIcon
+                  size="l"
+                  color="text"
+                  type={actionTypeRegistry.get(actionType.id).iconClass}
+                />
+              ),
+              isDisabled: !actionType.enabled,
+              onClick: () => setSelectedActionType(actionType),
+            }))}
+          />
+        ) : (
+          <EuiFlexGroup gutterSize="l">
+            {actionTypes?.map((actionType: ActionType) => (
```

Comment:
nit:

doesn't need the ?. here since the type is array here
should be wrapped in a useMemo. Since it's not possible here as it's after conditional I'd argue it's better to move the early return part with spinner out into the parent or make it as part of a HoC that this component is wrapped in before export.

Reply:
useMemo and key are fulfilling different levels of optimization, key is facilitating virtual dom reuse while useMemo facilitate react element creation, they work together. I thought we had the agreement to just memo everything because even tho individually it's not  that impressive on scale as a practice it adds up :| but up to you

Reply:
Yes, they are different things, but serve the same purpose, avoid updates.

If the component is memorized but we miss the key, React can not relate each item with its corresponding props (it uses the index as a fallback). This mismatch can potentially cause unnecessary updates to all the components in the array (If there's always the same number of items, in the same order the `key` is not strictly necessary, since using the `index` is enough)

With the `key` prop, React knows which is the relation between props and components, it will pass the correct props to each one. If the underlying component is memoized (it should be since we memoize everything), there's no need to do any extra memoization to the array.

Reply:
Sergi I understand how it works that's why I said they fullfil different levels of optimization which your response fully echoes so in that we are aligned. But respectfully my point was different:

if we have a convention should we follow it? Remember our yesterday discussion about context? it was your point to follow to point out to me to follow the convention even tho from perf perspective there was no immediate reason at all for that in that code.
Conversely If we don't have a convention, then disregard what I said. I was only under the assumption that we had that convention and as a reviewer I must raise the point.

About perf. If you ask me I believe there is no harm in memoizing everything. ANd that's exactly what I see in our codebase everywhere I look. But we can discuss it in eng sync again. feel free to resolve the discussion for now.w

Reply:
Those are very different topics, a React.context update can potentially cause many components to re-render, the benefit of memoizing only one place makes a big difference across the component tree. Here we are not even talking about re-renders, we are talking about memoizing small array loops. We have to use common sense.

You had things like this in your PR as well, and that's fine, I did not ask you to memoize every single small thing that could be memoized. I don't really understand what is the motivation behind all these comments.

Reply:
@semd you can find motivation underlined with red, I hope it helps to clarify the context
![image](https://github.com/user-attachments/assets/87c5ab89-6120-4c2b-b46b-a3f076ba459f)

> Here we are not even talking about re-renders, we are talking about memoizing small array loops.

It's not about small or big. If we have collectively decided that we are going to memoize everything, it includes small. It even includes TINY. Because on the scale of entire application memoizing many many tiny array calculations adds up as if a big one is memoized. And as react community learned a long time ago there is no such thing as too much memoize.

Btw to your point, If I had those things in my PR you should've told me that and I would fix them without objection. Moreover I'd say it was your duty as a reviewer to catch that when you saw, especially if you believe it's a convention. If it's not a convention though then you did nothing wrong so did I, but you seemingly refuse to say anything about that since the beginning of our conversation but rush to blame me for merely trying to uphold the convention if it exists as any reviewer should.

So now I am confused what is the motivation behind all your responses and why all that defensive attitude across these 2 threads ;)

Anyways... It's just code let's be friends we have a lot of PRs ahead of us we have to learn to communicate better and more intellectually honest with each other and assume good faith.

I'ill try again, have a nice weekend :)

Reply:
Hey @kapral18 , I didn't mean to sound defensive.
Let's rewind, the simple fact of the matter here is that we only have a small set of globally defined conventions in Kibana, and none of them talks about memoization strategies. Within the team, we've noticed issues that caused bad performance and historically we addressed them by memoizing components and props, that's why we encourage folks to "memoize everything" in PR reviews.
 
However, we have never formalized what "memoize everything" exactly means, we left that to interpretation and common sense, I guess. We need to actually write all of this knowledge down in a document so it's clearer for everyone. But, from what we have been doing for some time now, and from what seems to be the common practice in Kibana code, what I understand it means is: "memoize everything that may cause unnecessary re-renders". 

Scenarios like the ones we are talking about here have never been a matter of discussion, and I don't think they should.

I hope I have resolved your doubts, but don't hesitate to bring the topic to the team if you want 👍